### PR TITLE
Approve sdc by Guid

### PR DESCRIPTION
**Pull request by VamsiSiddu-7**
# Description
Add support for Approve SDC operation from CSI-PowerFlex driver

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| https://github.com/dell/csm/issues/402 |

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [x] Unit Testing (Refer GitHub workflow for more details) 
- [x] Functional Testing
      
  
preapproved-guid-node1

preapproved-guid-node2

### CODE REVIEW

File: service/node.go [hunk 1]
```diff
@@ -420,6 +428,33 @@ func (s *service) nodeProbe(ctx context.Context) error {
...
+		system := s.systems[systemID]
+
+		//fetch SDC details
+		sdc, err := s.systems[systemID].FindSdc("SdcGUID", opts.SdcGUID)
+		if err != nil {
+			return status.Errorf(codes.FailedPrecondition, "%s", err)
+		}
+
+		//fetch the restrictedSdcMode
+		if system.System.RestrictedSdcModeEnabled && system.System.RestrictedSdcMode == "Guid" {
```
Comment (JacobGros) [inline service/node.go:443]: We should log/return something if restricted mode is on, but the mode isn't supported by driver

Comment (JacobGros) [inline service/node.go:443]: Could we just check  system.System.RestrictedSdcMode here? 
I thought it was sent to None if the mode wasn't enabled. If I'm not remembering right, ignore this comment.

Comment (VamsiSiddu-7) [reply service/node.go:443]: yes you are right. if the RestrictedSdcModeEnabled is false the RestrictedSdcMode is set to "None".  I will update it.

File: service/service.go [hunk 2]
```diff
@@ -388,6 +390,13 @@ func (s *service) BeforeServe(
 	if sdcPrefix, ok := csictx.LookupEnv(ctx, EnvSDCPrefix); ok {
 		opts.SdcPrefix = sdcPrefix
 	}
+	if approveSDC, ok := csictx.LookupEnv(ctx, EnvIsApproveSDCEnabled); ok {
```
Comment (bharathsreekanth) [inline service/service.go:393]: Could simplify code, set opts.IsApproveSDCEnabled set to false by default and remove else clause. Upto you.

**Review by bharathsreekanth — APPROVED** [review]

File: service/features/get_system_instances.json [hunk 3]
```diff
@@ -151,8 +151,8 @@
     "swid": "",
     "defaultIsVolumeObfuscated": false,
     "daysInstalled": 6,
-    "restrictedSdcModeEnabled": false,
-    "restrictedSdcMode": "None",
+    "restrictedSdcModeEnabled": true,
```
[the remark this answers is not in the document]
Comment (VamsiSiddu-7) [reply service/features/get_system_instances.json:154]: yes Bharath

File: helm/csi-vxflexos/templates/node.yaml [hunk 4]
```diff
@@ -203,6 +203,10 @@ spec:
             - name: X_CSI_HEALTH_MONITOR_ENABLED
               value: "{{ .Values.node.healthMonitor.enabled }}"
             {{- end }}
+            {{- if eq .Values.node.approveSDCEnabled true}}
```
Comment (JacobGros) [inline helm/csi-vxflexos/templates/node.yaml:206]: should we check if this key exists first?
if it doesn't - set value to false or don't set at all

File: service/node.go (hunk 1 above)
Comment (VamsiSiddu-7) [reply service/node.go]: done

File: helm/csi-vxflexos/templates/node.yaml (hunk 4 above)
Comment (VamsiSiddu-7) [reply helm/csi-vxflexos/templates/node.yaml]: done

File: service/node.go [hunk 5]
```diff
@@ -420,6 +428,47 @@ func (s *service) nodeProbe(ctx context.Context) error {
 	return nil
 }
 
+func (s *service) approveSDC(opts Opts) error {
+
+	for _, systemID := range connectedSystemID {
+		system := s.systems[systemID]
+
+		if system == nil {
+			return nil
```
Comment (JacobGros) [inline service/node.go:437]: we shouldn't return here, but instead just continue to next systemID. 

Reason being: if there are two connected systems but first one is 00000..., second system should still have SDC approved

File: service/node.go [hunk 6]
```diff
@@ -420,6 +428,47 @@ func (s *service) nodeProbe(ctx context.Context) error {
...
+				}
+				Log.Infof("SDC Approved, SDC Id: %s and SDC GUID: %s", resp.SdcID, sdc.Sdc.SdcGUID)
+			} else {
+				Log.Infof("SDC already approved, SDC GUID: %s", sdc.Sdc.SdcGUID)
+			}
+		} else {
+			if !sdc.Sdc.SdcApproved {
+				return status.Errorf(codes.FailedPrecondition,
+					"Array RestrictedSdcMode is %s, driver only supports GUID RestrictedSdcMode cannot approve SDC %s",
+					system.System.RestrictedSdcMode, sdc.Sdc.SdcGUID)
```
Comment (JacobGros) [inline service/node.go:461]: is it possible to fix the spacing on this print line?

File: helm/csi-vxflexos/values.yaml [hunk 7]
```diff
@@ -209,6 +209,16 @@ node:
     # Default value: none
     # Examples: "rhel-sdc", "sdc-test"
     prefix: "sdc-test"
+   
+  # "approveSDC" defines the approve operation for SDC
+  # Default value: None
+  approveSDC: 
+    # enabled: Enable/Disable SDC approval
```
Comment (JacobGros) [inline helm/csi-vxflexos/values.yaml:216]: Can we make comment more descriptive?
maybe something like:
```
# enabled: Enable/Disable SDC approval`
#Allowed values:
#  true: Driver will attempt to approve restricted SDC by GUID during setup 
#  false: Driver will not attempt to approve restricted SDC by GUID during setup
# Default value: false
enabled: false    

``

File: service/node.go (hunk 5 above)
Comment (VamsiSiddu-7) [reply service/node.go]: done

File: service/node.go (hunk 6 above)
Comment (VamsiSiddu-7) [reply service/node.go:460]: done

File: helm/csi-vxflexos/values.yaml (hunk 7 above)
Comment (VamsiSiddu-7) [reply helm/csi-vxflexos/values.yaml:216]: done

**Review by JacobGros — APPROVED** [review]
LGTM

**Review by ashleyvjoy — DISMISSED** [review]

**Review by ashleyvjoy — APPROVED** [review]

**Review by JacobGros — APPROVED** [review]
LGTM
are these changes coming to operator as well?

**Review by bharathsreekanth — APPROVED** [review]

**Comment by VamsiSiddu-7** [comment]
> LGTM are these changes coming to operator as well?

yes jacob

**Pull request merged**